[WFLY-20393] Ensure that only one bytecode transformer is added for each Persistence Unit (also includes [WFLY-20394])

scottmarlow opened the pull request:
https://issues.redhat.com/browse/WFLY-20393
https://issues.redhat.com/browse/WFLY-20394 also was added which covers `Ensure that bytecode enhancement only happens if all persistence units allow bytecode enhancement`
____

<--- THIS SECTION IS AUTOMATICALLY GENERATED BY WILDFLY GITHUB BOT. ANY MANUAL CHANGES WILL BE LOST. --->

> Wildfly issue links:
> * [WFLY-20394](https://issues.redhat.com/browse/WFLY-20394)

<--- END OF WILDFLY GITHUB BOT REPORT --->

More information about the [wildfly-bot[bot]](https://github.com/wildfly/wildfly-github-bot)

jamezp commented on jpa/spi/src/main/java/org/jipijapa/plugin/spi/PersistenceUnitMetadata.java:
```diff
@@ -90,4 +90,12 @@ public interface PersistenceUnitMetadata extends PersistenceUnitInfo {
 
     List<ClassTransformer> getTransformers();
 
+    Object getLockObject();
```
We shouldn't use this. The `wasDelegatingTransformerAdded` and `setDelegatingTransformerAdded` should handle the locking and not require a consumer to know that.

scottmarlow replied:
Addressed in change just pushed.

jamezp commented on jpa/spi/src/main/java/org/jipijapa/plugin/spi/PersistenceUnitMetadata.java:
```diff
@@ -90,4 +90,12 @@ public interface PersistenceUnitMetadata extends PersistenceUnitInfo {
 
+    Object getLockObject();
+
+    /**
+      * @return true if a DelegatingClassTransformer was already setup to transform application classes for this persistence unit.
+     */
+    boolean wasDelegatingTransformerAdded();
+
+    void setDelegatingTransformerAdded();
```
IMO this should simply be a CAS. Something like:
```java
boolean compareAndSetDelegatingTransformerAdded(boolean expected, boolean value);
```

scottmarlow replied:
Switched to a compare and set, thanks!  And just pushed the change.

jamezp commented on jpa/subsystem/src/main/java/org/jboss/as/jpa/config/PersistenceUnitMetadataImpl.java:
```diff
@@ -100,6 +101,7 @@ public class PersistenceUnitMetadataImpl implements PersistenceUnitMetadata {
 
     private volatile Map<URL, Index> annotationIndex;
 
+    private volatile AtomicBoolean coordinateDelegatingTransformerAdd = new AtomicBoolean(false);
```
This should be final.

jamezp commented on jpa/subsystem/src/main/java/org/jboss/as/jpa/config/PersistenceUnitMetadataImpl.java:
```diff
@@ -381,11 +383,14 @@ public List<ClassTransformer> getTransformers() {
-                    classTransformer.getClass().getName(),
-                    getScopedPersistenceUnitName());
+        // WFLY-20393 Ensure that only one bytecode transformer is added for each Persistence Unit
+        if (true == coordinateDelegatingTransformerAdd.compareAndSet(false, true)) {
```
Not a big deal, but the `true ==` is redundant.

scottmarlow replied:
fixed

jamezp commented on jpa/subsystem/src/main/java/org/jboss/as/jpa/config/PersistenceUnitMetadataImpl.java:
```diff
@@ -381,11 +383,14 @@ public List<ClassTransformer> getTransformers() {
-                    classTransformer.getClass().getName(),
-                    getScopedPersistenceUnitName());
+        // WFLY-20393 Ensure that only one bytecode transformer is added for each Persistence Unit
+        if (true == coordinateDelegatingTransformerAdd.compareAndSet(false, true)) {
+            transformers.add(classTransformer);
+            if (ROOT_LOGGER.isTraceEnabled()) {
+                ROOT_LOGGER.tracef("added entity class transformer '%s' for '%s'",
+                        classTransformer.getClass().getName(),
+                        getScopedPersistenceUnitName());
+            }
```
Actually, I don't think this is correct. The `PersistenceUnitMetadata.getTransformers()` is a list so we should allow more than one I would think.

scottmarlow replied:
It looks right  as the Hibernate ORM/EclipseLink persistence provider is calling PersistenceUnitInfo#addTransformer(ClassTransformer classTransformer) to add a single `classTransformer` to our internal list of transformers.

jamezp replied:
That may be, but from the API it appears to allow for more than one so we're breaking the contract IMO.

I think "one time only" check needs to be on registering the `JPADelegatingClassFileTransformer` which is a different type of transformer. The latter being a JBoss Modules one, this one being a JPA one, e.g. `jakarta.persistence.spi.ClassTransformer`.

scottmarlow replied:
The API only allows one transformer to be added at a time but many transformers can be added by calling `addTransformer` multiple times which is still supported by the current code unless you see a problem that I don't :).

From https://jakarta.ee/specifications/persistence/3.1/apidocs/jakarta.persistence/jakarta/persistence/spi/persistenceunitinfo#addTransformer(jakarta.persistence.spi.ClassTransformer):

```
void addTransformer​(ClassTransformer transformer)

Add a transformer supplied by the provider that will be called for every new class definition or class redefinition that gets loaded by the loader returned by the getClassLoader() method. The transformer has no effect on the result returned by the getNewTempClassLoader() method. Classes are only transformed once within the same classloading scope, regardless of how many persistence units they may be a part of.

Parameters:
    transformer - provider-supplied transformer that the container invokes at class-(re)definition time 
```

Here is the transformer list from a few lines above in PersistenceUnitMetadataImpl.java:
```
// transformers will be written to when the Jakarta Persistence persistence provider adds their transformer.
    // there should be very few calls to add transformers but potentially many calls to get the
    // transformer list (once per application class loaded).
    private final List<ClassTransformer> transformers = new CopyOnWriteArrayList<ClassTransformer>();
```

scottmarlow replied:
Back to your first comment:

> Actually, I don't think this is correct. The PersistenceUnitMetadata.getTransformers() is a list so we should allow more than one I would think.

We cannot change the Persistence API which only allows one transformer to be added per `addTransformer` call but Hibernate call `addTransformer` multiple times,

jamezp replied:
That is what I'm saying :) We are changing the API for this. We're only allowing one to be added and we should not be doing that. In the `PersistenceUnitMetadataImple.addTransformer()` you're only allowing for one transformer to be added. Note this is a `jakarta.persistence.spi.ClassTransformer` _not_ a `org.jboss.modules.ClassTransformer`. We want to limit the number of times a `org.jboss.modules.ClassTransformer` is executed.

The issue is we're issuing the "has the transformer been added" check in the wrong place. It needs to happen in the `org.jboss.as.jpa.processor.JPAClassFileTransformerProcessor`. Because we may have multiple PU's, we need to have something like `PersistenceUnitMetadata.compareAndSetTransformersRegsiter(boolean, boolean)`.

scottmarlow replied:
Thanks James, I'm looking again at this!

scottmarlow replied:
Correction made!  :-)

Sanne commented on jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPAClassFileTransformerProcessor.java:
```diff
@@ -38,21 +41,26 @@ private void setClassLoaderTransformer(DeploymentUnit deploymentUnit) {
-                        if (Configuration.needClassFileTransformer(pu)) {
-                            transformer.addTransformer(new JPADelegatingClassFileTransformer(pu));
+                        if (!Configuration.needClassFileTransformer(pu)) {
+                            // WFLY-20394 Ensure that bytecode enhancement only happens if all persistence units allow bytecode enhancement
```
This comment sounds misleading? What if the current PU allows enhancement, but another doesn't?

scottmarlow replied:
For each application deployment, we are detecting a mix of enhanced + non-enhanced and will disable enhancement for all (EE container managed) persistence units in mixed case.

More concretely, if there are 160 persistence units and 159 have enhancement enabled but 1 doesn't, all 160 persistence units will not be enhanced.  We can revise this logic in the future as needed but IMO I think it matches what we agreed to in https://hibernate.zulipchat.com/#narrow/channel/132094-hibernate-orm-dev/topic/ORM.20bytecode.20enhancement.20.2B.20WildFly

@Sanne does ^ make sense?

scottmarlow replied:
I'm going to add a warning for the mixed case so users might be informed of it.

scottmarlow replied:
I made a change to show a warning and correct the check to ensure there is actually a mix of enhanced + not enhanced persistence units.  

Example warning: 
`WARN  [org.jboss.as.jpa] (MSC service thread 1-6) WFLYJPA0076: Some persistence units are configured to be enhanced but persistence unit(s) { 2lcmixedenhancement.jar#second_PU, 2lcmixedenhancement.jar#third_PU } disabled enhancement, none of the persistence units will be enhanced for this application deployment
`

Note that we show the list of persistence units that might need to be changed to allow enhancement so that all of the persistence units will be enhanced.

scottmarlow replied:
@Sanne I'm going to mark [this](https://github.com/wildfly/wildfly/pull/18757) feedback as resolved.  Please comment again if there are any concerns.

Sanne replied:
> More concretely, if there are 160 persistence units and 159 have enhancement enabled but 1 doesn't, all 160 persistence units will not be enhanced.

Sorry I tried to re-read the thread, but I still don't understand the rationale for disabling enhancement. When it's required, it doesn't seem safe to me to not enhance them. Either we do enhance them all and log a warning about this inconsistency, or you throw an error requiring the inconsistency to be addressed.

Clearly the warning would be less disruptive for existing applications so I expect that is what we'd do here, but perhaps in the future it's worth making it stricter as the side-effects can be rather tricky.

scottmarlow replied:
Addressed as per discussion today.

scottmarlow committed:
[WFLY-20393] Ensure that only one transformer is added for each Persistence Unit

Signed-off-by: Scott Marlow <smarlow@redhat.com>

bstansberry commented on jpa/subsystem/src/main/java/org/jboss/as/jpa/messages/JpaLogger.java:
```diff
@@ -757,4 +757,14 @@ public interface JpaLogger extends BasicLogger {
+     * @param scopedPuName name of the persistence unit
+     */
+    @LogMessage(level = WARN)
+    @Message(id = 76, value = "Some persistence units are configured to be enhanced but persistence unit(s) { %s } disabled enhancement, none of the persistence units will be enhanced for this application deployment")
```
s/enhancement, none/enhancement; therefore none/g

bstansberry commented on jpa/subsystem/src/main/java/org/jboss/as/jpa/messages/JpaLogger.java:
```diff
@@ -757,4 +757,14 @@ public interface JpaLogger extends BasicLogger {
+     * Logs warning detected a mix of bytecode enhanced + non-enhanced, will disable enhancement for all persistence units in this deployment.
+     * Also see discussion on https://hibernate.zulipchat.com/#narrow/channel/132094-hibernate-orm-dev/topic/ORM.20bytecode.20enhancement.20.2B.20WildFly if possible.
+     *
+     * @param scopedPuName name of the persistence unit
```
I think the param should be scopedPuNames. So then

@param scopedPuNames comma-delimited names of the persistence units

scottmarlow replied:
Change made locally and will be included in next push.

bstansberry commented on jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPAClassFileTransformerProcessor.java:
```diff
@@ -38,21 +44,37 @@ private void setClassLoaderTransformer(DeploymentUnit deploymentUnit) {
                 }
             }
+
+            if (notEnhancedPersistenceUnits.size() > 0 && persistentUnits.size() > 0 ) {
```
This guard will return true in the case where *no* PU has enhancement enabled. There shouldn't be a WARN in such a case.

It might be cleaner to rename 'persistentUnits' to 'enhancedPersistenceUnits' and use an if/else for L55-58. The if test here would then only be true if both sets have entries.

scottmarlow replied:
Hmm, already updated locally to the following (single loop and using if/else), will look for the ^ problem in :

```
    private void setClassLoaderTransformer(DeploymentUnit deploymentUnit) {
        // (AS7-2233) each persistence unit can use a persistence provider, that might need
        // to use ClassTransformers.  Providers that need class transformers will add them
        // during the call to CreateContainerEntityManagerFactory.

        DelegatingClassTransformer transformer = deploymentUnit.getAttachment(DelegatingClassTransformer.ATTACHMENT_KEY);
        if ( transformer != null) {
            final List<PersistenceUnitMetadata> notEnhancedPersistenceUnits = new ArrayList<>();
            int enhancedPersistenceUnitsCount = 0;
            for (ResourceRoot resourceRoot : DeploymentUtils.allResourceRoots(deploymentUnit)) {
                PersistenceUnitMetadataHolder holder = resourceRoot.getAttachment(PersistenceUnitMetadataHolder.PERSISTENCE_UNITS);
                if (holder != null) {
                    for (PersistenceUnitMetadata pu : holder.getPersistenceUnits()) {
                        if (!Configuration.needClassFileTransformer(pu)) {
                            notEnhancedPersistenceUnits.add(pu);
                        } else if ( pu.needsJPADelegatingClassFileTransformer()) {
                            enhancedPersistenceUnitsCount++;
                            transformer.addTransformer(new JPADelegatingClassFileTransformer(pu));
                        }
                    }
                }
            }

            if (notEnhancedPersistenceUnits.size() > 0 && enhancedPersistenceUnitsCount > 0 ) {
                // WFLY-20394 Log warning if applications have a mix of some persistence units with "jboss.as.jpa.classtransformer" set to false and some set to true
                String persistenceUnitsConfiguredNotEnhanced = notEnhancedPersistenceUnits.
                        stream().map(PersistenceUnitMetadata::getScopedPersistenceUnitName).
                        collect(Collectors.joining(", "));
                ROOT_LOGGER.mixedEnhancedAndNotEnhanced(persistenceUnitsConfiguredNotEnhanced);
            }
       }
    }
```

bstansberry reviewed (changes requested):
Beyond my comments below, this WARN behavior seems like a good thing to cover in an integration test. Testing for the presence of log statements is a bit fussy, but we have testsuite utilities that help.

Sanne commented:
I had added another comment, but realizing it's on a "resolved concern" it's not made visible by github, so calling it out here.
  - https://github.com/wildfly/wildfly/pull/18757#discussion_r1960570509

scottmarlow committed:
[WFLY-20394] Log warning if applications have a mix of some persistence units with "jboss.as.jpa.classtransformer" set to false and some set to true

Signed-off-by: Scott Marlow <smarlow@redhat.com>

scottmarlow commented:
> I had added another comment, but realizing it's on a "resolved concern" it's not made visible by github, so calling it out here.
> 
>     * [[WFLY-20393] Ensure that only one bytecode transformer is added for each Persistence Unit (also includes [WFLY-20394]) #18757 (comment)](https://github.com/wildfly/wildfly/pull/18757#discussion_r1960570509)

Thank you Sanne!  The linked concern was updated to be visible and should be now addressed by change to only show the warning.

scottmarlow commented:
> testsuite utilities

Is https://github.com/wildfly/wildfly/blob/main/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/messaging/mgmt/NoAuditLogTestCase.java#L61 an example of reading the log that we should follow here?

jamezp commented on jpa/subsystem/src/main/java/org/jboss/as/jpa/processor/JPAClassFileTransformerProcessor.java:
```diff
@@ -40,19 +46,30 @@ private void setClassLoaderTransformer(DeploymentUnit deploymentUnit) {
+                        if (!Configuration.needClassFileTransformer(pu)) {
+                            notEnhancedPersistenceUnits.add(pu);
+                        } else if ( pu.needsJPADelegatingClassFileTransformer()) {
+                            enhancedPersistenceUnitsCount++;
```
Not critical, but this could be a simple `boolean`.

bstansberry commented:
> > Beyond my comments below, this WARN behavior seems like a good thing to cover in an integration test. Testing for the presence of log statements is a bit fussy, but we have testsuite utilities that help.
> 
> Thanks @bstansberry, I'll look to use wildfly/testsuite /shared/src/main/java/org/jboss/as/test/shared/util/LoggingUtil.java

@scottmarlow See also org.jboss.as.test.shared.TestLogHandlerSetupTask. 

There's similar stuff in org.wildfly.core:wildfly-core-testsuite-shared too: org.jboss.as.test.shared.logging.LoggingUtil and org.jboss.as.test.shared.logging.TestLogHandlerSetupTask. But WF Core doesn't use Arquillian so those aren't adapted to the WildFly Arquillian APIs.

scottmarlow commented:
> > > Beyond my comments below, this WARN behavior seems like a good thing to cover in an integration test. Testing for the presence of log statements is a bit fussy, but we have testsuite utilities that help.
> > 
> > 
> > Thanks @bstansberry, I'll look to use wildfly/testsuite /shared/src/main/java/org/jboss/as/test/shared/util/LoggingUtil.java
> 
> @scottmarlow See also org.jboss.as.test.shared.TestLogHandlerSetupTask.
> 
> There's similar stuff in org.wildfly.core:wildfly-core-testsuite-shared too: org.jboss.as.test.shared.logging.LoggingUtil and org.jboss.as.test.shared.logging.TestLogHandlerSetupTask. But WF Core doesn't use Arquillian so those aren't adapted to the WildFly Arquillian APIs.

Actually, it might make sense to not log a warning but use TRACE logging.  Will think some more about this.

scottmarlow commented:
> Should we perhaps separate the component upgrades from the rest of this? If they work properly without the other changes, then there is value in getting them in and getting the bake from all the CI we run.

@bstansberry 
https://github.com/wildfly/wildfly/pull/18748 only contains the component upgrades now.

https://github.com/scottmarlow/wildfly/tree/WFLY-20393_addonetransformer_per_persistenceunit has the other TBD changes now that corresponds to this pr.  Will see which of the `WFLY-20393_addonetransformer_per_persistenceunit` branch changes we need.
